### ISSUE TITLE
PiP is not working 

### ISSUE THREAD

**Issue by loverboy0004** [CypherpunkArmory/UserLAnd#473]
i try many time with 
“pip install flask”
it gives memory error
try with --no-cache gives error again
then try with python
“python -m pip install flask” 
error continues again and again 
also try with pip2 and easy_install 
then try to reset the software and also try with python3 pip3
Now you tell me how i solve it

**Comment by foodmaher** [comment]
@loverboy0004 ,
I did the first thing on Termux directly ...
` pip install flask `

**Comment by loverboy0004** [comment]
bro i was using GUNRoot no Problem with that
but GunRoot is not supporting Android 9
so im new in UserLand and try to solve the issue with pip but i failed
again and again
then i think i connect the maker of this software
check this image

**Comment by loverboy0004** [comment]
and im noticing now npm is also giving error not installing software
please fix GUNRoot i dont like UserLand
so many bugs in it please Update your software and fix the bugs

**Comment by foodmaher** [comment]
@loverboy0004 ,
My brother, I am not part of the development team. I just tried to check it out and work with me on Termux and not Userland .... 😘😘😘😘😘
But I'll try it on the Ubuntu distro in Userland and tell you what I'm going to get ...

      Note
My Driver : LG G4 marshmallow 6.0
Processor : Snapdragon 808 aarch64 
build : 2015 year

**Comment by loverboy0004** [comment]
sorry bro but did you try GUNROOT its best but sadly its still not
supporting android oreo :(
Gunroot is 10 time batter then termux im using Gunroot over 2 years

**Comment by MatthewTighe** [comment]
@loverboy0004 Any bugs you're experiencing would be present in GNURoot now if we had continued to support it. The two projects use the same underlying PRoot code, but we needed a more maintainable and modern codebase for Android-facing features. 

Here is our bug report template:
```
Describe the bug

A clear and concise description of what the bug is.

To Reproduce

Steps to reproduce the behavior:

Go to '...'
Click on '....'
Scroll down to '....'
See error
Expected behavior

A clear and concise description of what you expected to happen.

Screenshots

If applicable, add screenshots to help explain your problem.

Device Information

Device: [e.g. Pixel XL 2]
Android Version: [e.g. Oreo, 8.0]
UserLAnd Version [e.g. 0.1.2]
Additional context

Add any other context about the problem here.
```

**Comment by casept** [comment]
I'm experiencing this bug as well.
Issue template follows:

Describe the bug:

pip (both pip2 and pip3) fail to install any packages on Ubuntu. Both the `pip` CLI  tool as well as trying to run `import pip` in the python(3) REPL result in the following stacktrace:

```
user@localhost:~$ pip3
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/_vendor/__init__.py", line 33, in vendored
    __import__(vendored_name, globals(), locals(), level=0)
ModuleNotFoundError: No module named 'pip._vendor.colorama'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/pip3", line 9, in <module>
    from pip import main
  File "/usr/lib/python3/dist-packages/pip/__init__.py", line 21, in <module>
    from pip._vendor.requests.packages.urllib3.exceptions import DependencyWarning
  File "/usr/lib/python3/dist-packages/pip/_vendor/__init__.py", line 65, in <module>
    vendored("colorama")
  File "/usr/lib/python3/dist-packages/pip/_vendor/__init__.py", line 36, in vendored
    __import__(modulename, globals(), locals(), level=0)
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/usr/share/python-wheels/colorama-0.3.7-py2.py3-none-any.whl/colorama/__init__.py", line 2, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/usr/share/python-wheels/colorama-0.3.7-py2.py3-none-any.whl/colorama/initialise.py", line 6, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/usr/share/python-wheels/colorama-0.3.7-py2.py3-none-any.whl/colorama/ansitowin32.py", line 7, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/usr/share/python-wheels/colorama-0.3.7-py2.py3-none-any.whl/colorama/winterm.py", line 2, in <module>
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 656, in _load_unlocked
  File "<frozen importlib._bootstrap>", line 626, in _load_backward_compatible
  File "/usr/share/python-wheels/colorama-0.3.7-py2.py3-none-any.whl/colorama/win32.py", line 8, in <module>
  File "/usr/lib/python3.6/ctypes/__init__.py", line 538, in <module>
    _reset_cache()
  File "/usr/lib/python3.6/ctypes/__init__.py", line 273, in _reset_cache
    CFUNCTYPE(c_int)(lambda: None)
MemoryError
```

Steps to reproduce the behaviour:
1. Create a fresh Ubuntu roots
2. Install python and pip using `sudo apt update && sudo apt install python3-pip`
3. Try to install a package, for example `pip3 install pipenv`
4. Observe crash

Expected behavior: pip installing the package

Device information: 
* Device: Galaxy S5 (SM-G900F)
* Android version: 8.1 (LineageOS 14.1)
* UserLAnd version: 1.0.3 (F-Droid)

Additional context:
Removing line 273 from `/use/lib/python3.6/ctypes/__init__.py` seems to work around the bug.

**Comment by loverboy0004** [comment]
thanks alot i reinstall the software try again

**Comment by loverboy0004** [comment]
i reinstall software again
and try with Ubuntu
pip not working
then try with kali and then try with debian
pip and npm not working
i need them i think i have to find the new software for debian in android

**Comment by loverboy0004** [comment]
im using Samsung Galaxy C5
with android v9.0
GUNROOT is not supporting this version otherwise i m using only that software

**Comment by corbinlc** [comment]
These sound like a real bug and I should be able to start looking at them soon.

**Comment by loverboy0004** [comment]
@corbinlc i also try get_pip same error , try easy_install and also try to install pipenv same error .. you fix this bug as soon as possible with in days but for now i need a package like yarn in nodejs because its working and npm is also not working at all please for my project i need pip now please tell me any package which replaces pip

**Comment by loverboy0004** [comment]
@stephenprater bro you tell me i Manually edit that file or reinstall colorama

**Comment by stephenprater** [comment]
Manually edit the file.   Just remove the call to `init` altogther.

Like this:

```
@contextlib.contextmanager
def colorama_text(*args, **kwargs):
    try:
        yield
    finally:
        deinit()
```

You might need to do a little experimenting to get it to work.  The point is to never hit the `import ctype` on line 8 of `win32.py` which is where the error raised.

**Comment by loverboy0004** [comment]
not able to edit bro
error no such a file or folder
/usr/share/python-wheels/colorama-0.3.7-py2.py3-none-any.whl/colorama
it shows colorama-0.3.7-py2.py3-none-any.whl is not a folder

**Comment by loverboy0004** [comment]
finally i find the file
usr/lib/python2.7/dist-packages/colorama/initialise.py
i edit it like you tell me but no effects
i try to reinstall but error comes


root@localhost:/colorama-0.3.3# python setup.py install
Traceback (most recent call last):
  File "setup.py", line 9, in <module>
    from setuptools import setup
  File "/usr/lib/python2.7/dist-packages/setuptools/__init__.py", l
ine 14, in <module>
    from setuptools.dist import Distribution, Feature
  File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line
25, in <module>
    from setuptools import windows_support
  File "/usr/lib/python2.7/dist-packages/setuptools/windows_support
.py", line 2, in <module>
    import ctypes
  File "/usr/lib/python2.7/ctypes/__init__.py", line 556, in <modul
e>
    _reset_cache()
  File "/usr/lib/python2.7/ctypes/__init__.py", line 276, in _reset
_cache
    CFUNCTYPE(c_int)(lambda: None)
MemoryError

**Comment by loverboy0004** [comment]
root@localhost:/colorama-0.3.3# python setup.py install
Traceback (most recent call last):
  File "setup.py", line 9, in <module>
    from setuptools import setup
  File "/usr/lib/python2.7/dist-packages/setuptools/__init__.py", l
ine 14, in <module>
    from setuptools.dist import Distribution, Feature
  File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line
25, in <module>
    from setuptools import windows_support
  File "/usr/lib/python2.7/dist-packages/setuptools/windows_support
.py", line 2, in <module>
    import ctypes
  File "/usr/lib/python2.7/ctypes/__init__.py", line 556, in <modul
e>
    _reset_cache()
  File "/usr/lib/python2.7/ctypes/__init__.py", line 276, in _reset
_cache
    CFUNCTYPE(c_int)(lambda: None)
MemoryError

**Comment by loverboy0004** [comment]
@stephenprater i try to reinstall python3.7 
with Ubuntu PPA
but still same error when i type python3.7 -m pip --version
any suggestion may be we solve the error with installing the corrupt package with PPA ???

**Comment by corbinlc** [comment]
`pip install flask` works fine on a Pixel running 9.0.  Looks like this is a device specific issue.  Both of you have Samsung devices, so I am guessing it has to do with many files under `/proc` not being readable on Samsung devices running 8.0+.  The way we have handled this is by emulating the missing file.  I will try this on a Galaxy S9 and see if the problem shows up there.

**Comment by corbinlc** [comment]
This flow works both on a Pixel and a Galaxy S9 running Android 9.0 from within an Ubuntu filesystem.
```
sudo apt update
sudo apt install python-pip
pip install flash
```
Do you @casept and @loverboy0004 expect that should have failed?  Or is the recipe for getting to the failure more complicated?
Can you guys provide all the commands to get to the reported error starting from a new Ubuntu filesystem?  Looks like I am missing something.

**Comment by corbinlc** [comment]
I notice that you guys said that `pip` and `pip3` both fail, but the example provided is `pip3` . Trying that now.

**Comment by corbinlc** [comment]
Following this: 

Steps to reproduce the behavior:
1. Create a fresh Ubuntu roots
2. Install python and pip using `sudo apt update && sudo apt install python3-pip`
3. Try to install a package, for example `pip3 install pipenv`

I get no crash at all.  I only get success.
I notice that @casept is using LineageOS, are you using LineageOS as well @loverboy0004?  That might be a difference that matters.

**Comment by corbinlc** [comment]
Btw... according to your reviews (both of them) that you have left on Google Play, you are using Android 8.0 and not Android 9 @loverboy0004 .

**Comment by corbinlc** [comment]
Just a quick rant while I am waiting on the next piece of data.  @loverboy0004 you have provided two google play reviews, you have emailed us at two different email addresses, you have filed duplicate issues on this github project and you have filed an issue against GNURoot.  This is not helpful and is instead just obnoxious.  This one issue is sufficient, but you could have provided more info, like @casept did.  That would have been helpful.  I will continue to debug this because it is interesting to me and probably others.  While you are waiting for someone else to solve a problem you are having with a tool that you got freely, I suggest you chill out.

**Comment by loverboy0004** [comment]
@corbinlc sorry i have no option .. I'm tring to fix it .. you don't know i completely lost my project .. because i need dryscrape for my project but because of pip i can't able to install it .. only that's why i over reacted 
now i tell you about my research about pip
--------------------------See Blow------------------------
1st of all when i post this issue then i don't know this is not a bug on your software so sorry about that but when i try Ubuntu-on-termux then install pip and error comes i shocked that why ?? 

then i just try debian-on-termux but same error comes 

then i try to reinstall python with Ubuntu PPA/DeadSnake but pip and pip3 gives same error 

i try to install ctype and many packages but issues remains then i try to find any package because npm installation not working but yarn is working so i think i need a package like yarn then i found pipenv but i can't able to install it because when i try to type python install.py then MemoryError comes again 
---------------------------END-------------------------
that's all i tried till now but my project is useless til now but this Problem is only on Android 8+ with Samsung and some other models i thing they restrict something that's why problem comes 
but some other versions of linux pip is working very well i try fedora pip is working and also termux pip is working on termux 
im using :-
        SAMSUNG C5 
        ANDROID v8.0
        SAMSUNG EXPERIENCES v9.0
PiP is working on :-
       Termux
       Fedora-on-Termux
       Fedora on Userland (not tested yet)
PiP is not working on :-
       Ubuntu on termux 
       Ubuntu on Userland 
       Kali on Userland 
       Debian on Userland 
       Debian on Termux
and may be some others 

Now You turn Brother Thanks for listening me Thanks for being Here You guys are great 
Sorry Again

**Comment by corbinlc** [comment]
A few things:
1) Ubuntu on Termux uses PRoot like we do and our PRoot code is very similar to theirs.  So, I am not too surprised that it might not work there.  
2) Can you tell me whether you are using LineageOS or some other non-stock setup?
3) Can you do the following.  A) Within UserLAnd, stop any running apps. B) Within UserLAnd, go to settings (three dots in top right) and enable PRoot debug logging and set the debug log level to 9.  C) Recreate the failure a couple of times in a row.  D) You should find a new file called PRoot_Debug_Log on your sdcard (real or emulated).  I need you to provide that log to me.

**Comment by loverboy0004** [comment]
follow all steps but PRoot_Debug_Log not found 
try twice

**Comment by corbinlc** [comment]
Please see the question about LineageOS or another non-stock setup

**Comment by corbinlc** [comment]
On pretty much all devices the file is stored at /sdcard/PRoot_Debug_Log .  It is only create if you 1) enable logging and 2) set the verbosity to some value (please use 9)

**Comment by loverboy0004** [comment]
i don't know what is LineageOS 
------------------See Blow---------------
i open userland then i try you 1st step 
then i enable the PRoot_Debug_Log in setting 
then i set lvl 9
then i type pip install flask --verbose twice 
then file the  PRoot_Debug_Log file but cant able to find it 
-------------------END----------------
i missed something ??

**Comment by corbinlc** [comment]
On a Samsung device, if you go into your "My Files" app that Samsung puts on your device by default.  You can go to `My Files > Internal Storage` and find it there.  At least that is true on my Samsung S9.  If it is not there, please browse around a little.

**Comment by loverboy0004** [comment]
yes i know but no file there
also i search no results i try to find it on linux

**Comment by corbinlc** [comment]
I have never heard of the file not being created.  You have to start the app from scratch after setting up the PRoot logging settings for them to take effect.  Did you possible start the app before changing the settings?

**Comment by loverboy0004** [comment]
bro file will be on mnt/sdcard right ???
but when i try to open mnt/sdcard on linux it give error no such file or folder 
but when i type cd / and cd sdcard its open my sdcard

**Comment by loverboy0004** [comment]
i restarte the app and now i find the file how i send you that ?? copy paste  or email

**Comment by corbinlc** [comment]
Email or a pastebin or dropbox link or similar.  You choose.

**Comment by loverboy0004** [comment]
email is easy for me lol give me your email

**Comment by corbinlc** [comment]
You can use the email you used last time...
support@userland.tech

**Comment by loverboy0004** [comment]
Done bro

**Comment by corbinlc** [comment]
Another thing to try would be setting `TMPDIR` to some location that is known to allow execution when calling `pip3`.  For example, can you try this... `TMPDIR=/home/harry pip3 install pipenv`  (I saw your username was harry in the log you sent)

**Comment by loverboy0004** [comment]
TMPDIR is also not working and bro i email you __init__.py tell me which line i remove

**Comment by corbinlc** [comment]
You can see what file and what line from the error message you provided...

> root@localhost:/colorama-0.3.3# python setup.py install
> Traceback (most recent call last):
>   File "setup.py", line 9, in <module>
>     from setuptools import setup
>   File "/usr/lib/python2.7/dist-packages/setuptools/__init__.py", l
> ine 14, in <module>
>     from setuptools.dist import Distribution, Feature
>   File "/usr/lib/python2.7/dist-packages/setuptools/dist.py", line
> 25, in <module>
>     from setuptools import windows_support
>   File "/usr/lib/python2.7/dist-packages/setuptools/windows_support
> .py", line 2, in <module>
>     import ctypes
>   File "/usr/lib/python2.7/ctypes/__init__.py", line 556, in <modul
> e>
>     _reset_cache()
>   File "/usr/lib/python2.7/ctypes/__init__.py", line 276, in _reset
> _cache
>     CFUNCTYPE(c_int)(lambda: None)
> MemoryError

It says...
> File "/usr/lib/python2.7/ctypes/__init__.py", line 276

**Comment by corbinlc** [comment]
I have confirmed from the file you sent me that it is exactly that line that should be commented out or removed.
The line looks like this `CFUNCTYPE(c_int)(lambda: None)`

**Comment by corbinlc** [comment]
Btw... I am still unable to reproduce your error with the devices we have and the proot log does not look very enlightening.  A workaround like commenting out that line, is probably the best we can do for a while.

**Comment by loverboy0004** [comment]
well well well 
finally after more then 15 day 
so simple and so easy bug fixed thanks bro im now trying to install and checking up pip is working now thanks alot :)
but in future please update this app with built in terminal with many windows or tabs thanks you are very nice person

**Comment by corbinlc** [comment]
Glad it is working for you.  We are very close to releasing a built-in terminal.  If you are happy with the work we have done, please update your reviews now and in the future as we add more of the features you want.

**Comment by loverboy0004** [comment]
why not

**Comment by jwoehr** [comment]
> Have you tried doing what @casept suggested...
> 
> > Removing line 273 from /usr/lib/python3.6/ctypes/**init**.py seems to work around the bug.
> 
> Specifically that line looks like this...
> `CFUNCTYPE(c_int)(lambda: None)`
> 
> You can comment it out instead if desired....that is the file if dealing with pip3.

Thanks @corbinlc ... at this late date ran into same bug and your advice worked to get `pip3` working.

**Comment by yangruihan** [comment]
> Have you tried doing what @casept suggested...
> 
> > Removing line 273 from /usr/lib/python3.6/ctypes/**init**.py seems to work around the bug.
> 
> Specifically that line looks like this...
> `CFUNCTYPE(c_int)(lambda: None)`
> 
> You can comment it out instead if desired....that is the file if dealing with pip3.

Thanks, I got the same problem, and resolve it with commenting the `CFUNCTYPE(c_int)(lambda: None)`. 
:-) 👍
